File: src/repositories/attendanceRepository.js
// ---------------------------------------------------------------------
// ATTENDANCE REPOSITORY (Supabase only)
// Handles driver attendance data
// ---------------------------------------------------------------------

import { BaseRepository } from './baseRepository'
import supabase from '../lib/supabase'

const ATTENDANCE_SELECT = 'id, driver_name, driver_id, attendance_date, status, check_in, check_out, created_at, updated_at'
const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i

function toDbAttendance(data = {}) {
  const payload = {
    driver_id: UUID_RE.test(String(data.driver_id || '')) ? data.driver_id : undefined,
    driver_name: data.driver_name ?? data.driver,
    attendance_date: data.attendance_date ?? data.date,
    status: data.status,
    check_in: data.check_in ?? data.checkIn,
    check_out: data.check_out ?? data.checkOut,
  }
  return Object.fromEntries(Object.entries(payload).filter(([, value]) => value !== undefined))
}

/**
 * AttendanceRepository - Manages attendance data via Supabase
 */
export class AttendanceRepository extends BaseRepository {
  constructor() {
    super('attendance', 'id')
  }

  async getAll() {
    try {
      const { data, error } = await supabase
        .from('attendance')
        .select(ATTENDANCE_SELECT)
        .order('attendance_date', { ascending: false })
      if (error) throw error
      return data || []
    } catch (error) {
      console.error(error)
      throw error
    }
  }

  async getByDriver(driverId) {
    try {
      const { data, error } = await supabase
        .from('attendance')
        .select(ATTENDANCE_SELECT)
        .eq('driver_id', driverId)
      if (error) throw error
      return data || []
    } catch (error) {
      console.error(error)
      throw error
    }
  }

  async getByDate(date) {
    try {
      const { data, error } = await supabase
        .from('attendance')
        .select(ATTENDANCE_SELECT)
        .eq('attendance_date', date)
      if (error) throw error
      return data || []
    } catch (error) {
      console.error(error)
      throw error
    }
  }

  async getByMonth(month, year) {
    try {
      const startDate = `${year}-${String(month).padStart(2, '0')}-01`
      const endDate = `${year}-${String(month).padStart(2, '0')}-31`
      const { data, error } = await supabase
        .from('attendance')
        .select(ATTENDANCE_SELECT)
        .gte('attendance_date', startDate)
        .lte('attendance_date', endDate)
      if (error) throw error
      return data || []
    } catch (error) {
      console.error(error)
      throw error
    }
  }

  async create(data) {
    try {
      const record = {
        ...toDbAttendance(data),
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      }
      const { data: created, error } = await supabase
        .from('attendance')
        .insert([record])
        .select()
        .single()
      if (error) throw error
      return created
    } catch (error) {
      console.error(error)
      throw error
    }
  }

  async update(id, data) {
    try {
      const { data: updated, error } = await supabase
        .from('attendance')
        .update({
          ...toDbAttendance(data),
          updated_at: new Date().toISOString(),
        })
        .eq('id', id)
        .select()
        .single()
      if (error) throw error
      return updated
    } catch (error) {
      console.error(error)
      throw error
    }
  }

  async delete(id) {
    try {
      const { error } = await supabase
        .from('attendance')
        .delete()
        .eq('id', id)
      if (error) throw error
      return true
    } catch (error) {
      console.error(error)
      throw error
    }
  }
}

export const attendanceRepository = new AttendanceRepository()
export default AttendanceRepository

File: mobile/lib/repositories/attendance_repository.dart
// ─────────────────────────────────────────────────────────────────────────────
// attendance_repository.dart
// Supabase data layer for attendance.
//
// IMPORTANT — matching Web ERP data format:
//   check_in / check_out stored as TEXT "HH:MM" (IST local time)
//   working_hours stored as TEXT "X hours Y minutes"
//   attendance_date stored as DATE "YYYY-MM-DD" (local date)
//
// RLS policies enforced server-side:
//   attendance_driver_own_read   → SELECT
//   attendance_driver_own_insert → INSERT
//   attendance_driver_own_update → UPDATE (added Day 45)
// ─────────────────────────────────────────────────────────────────────────────

import 'package:supabase_flutter/supabase_flutter.dart' hide AuthState;
import '../core/config/supabase_config.dart';
import '../models/attendance.dart';

class AttendanceRepository {
  AttendanceRepository(this._client);

  final SupabaseClient _client;

  // ── Today ──────────────────────────────────────────────────────────────────

  /// Returns today's attendance record for the authenticated driver.
  /// Returns null if no record exists yet.
  /// RLS ensures driver can only get their own record.
  Future<AttendanceModel?> getTodayAttendance() async {
    final today = _todayString();
    final data = await _client
        .from(SupabaseConfig.attendanceTable)
        .select(AttendanceModel.selectColumns)
        .eq('attendance_date', today)
        .maybeSingle();

    if (data == null) return null;
    return AttendanceModel.fromMap(data);
  }

  // ── History ────────────────────────────────────────────────────────────────

  /// Last [limit] attendance records for the driver, newest first.
  Future<List<AttendanceModel>> getAttendanceHistory({int limit = 30}) async {
    final data = await _client
        .from(SupabaseConfig.attendanceTable)
        .select(AttendanceModel.selectColumns)
        .order('attendance_date', ascending: false)
        .limit(limit);

    return (data as List).map((m) => AttendanceModel.fromMap(m)).toList();
  }

  // ── Check-in ───────────────────────────────────────────────────────────────

  /// Creates a new attendance record for today.
  /// Throws if a record already exists (duplicate check enforced by caller).
  Future<AttendanceModel> checkIn({
    required String driverId,
    required String driverName,
  }) async {
    final now   = DateTime.now();
    final today = _todayString();
    final time  = _timeString(now);    // "HH:MM"

    final data = await _client
        .from(SupabaseConfig.attendanceTable)
        .insert({
          'driver_id':       driverId,
          'driver_name':     driverName,
          'attendance_date': today,
          'status':          'present',
          'check_in':        time,
          'notes':           'Checked in from mobile app',
        })
        .select(AttendanceModel.selectColumns)
        .single();

    return AttendanceModel.fromMap(data);
  }

  // ── Check-out ──────────────────────────────────────────────────────────────

  /// Updates existing attendance record with check-out time and worked hours.
  Future<AttendanceModel> checkOut({
    required String attendanceId,
    required String checkInTime,   // "HH:MM" from the existing record
  }) async {
    final now      = DateTime.now();
    final checkOut = _timeString(now);
    final hours    = _calcWorkingHours(checkInTime, checkOut);

    final data = await _client
        .from(SupabaseConfig.attendanceTable)
        .update({
          'check_out':     checkOut,
          'working_hours': hours,
          'updated_at':    now.toUtc().toIso8601String(),
        })
        .eq('id', attendanceId)
        .select(AttendanceModel.selectColumns)
        .single();

    return AttendanceModel.fromMap(data);
  }

  // ── Helpers ───────────────────────────────────────────────────────────────

  /// Today's date as "YYYY-MM-DD" in local time.
  String _todayString() {
    final d = DateTime.now();
    return '${d.year}-${d.month.toString().padLeft(2, '0')}-'
        '${d.day.toString().padLeft(2, '0')}';
  }

  /// Format DateTime as "HH:MM" local time — matches Web ERP format.
  String _timeString(DateTime dt) =>
      '${dt.hour.toString().padLeft(2, '0')}:'
      '${dt.minute.toString().padLeft(2, '0')}';

  /// Calculate working hours as TEXT string matching Web ERP format.
  /// e.g. "9 hours" or "8 hours 30 minutes"
  String _calcWorkingHours(String checkIn, String checkOut) {
    try {
      final inParts  = checkIn.split(':');
      final outParts = checkOut.split(':');

      final inMin  = int.parse(inParts[0])  * 60 + int.parse(inParts[1]);
      final outMin = int.parse(outParts[0]) * 60 + int.parse(outParts[1]);

      final diff    = outMin - inMin;
      if (diff <= 0) return '0 hours';

      final hours   = diff ~/ 60;
      final minutes = diff % 60;

      if (minutes == 0) return '$hours ${hours == 1 ? 'hour' : 'hours'}';
      return '$hours ${hours == 1 ? 'hour' : 'hours'} $minutes minutes';
    } catch (_) {
      return 'N/A';
    }
  }
}
